unstructured/documents/elements.py

```diff
@@ -772,20 +786,61 @@ def __init__(
         self.embeddings: Optional[list[float]] = embeddings
 
         if isinstance(element_id, NoID):
-            # NOTE(robinson) - Cut the SHA256 hex in half to get the first 128 bits
-            element_id = hashlib.sha256(text.encode()).hexdigest()[:32]
-
+            self._id = self.calculate_hash(metadata)
         elif isinstance(element_id, UUID):
-            element_id = str(uuid.uuid4())
+            self._id = uuid.uuid4()
+        elif isinstance(element_id, str):
+            self._id = element_id
+        else:
+            raise ValueError("ID must be a string, UUID, or NoID")
 
         super().__init__(
-            element_id=element_id,
+            element_id=self._id,
             metadata=metadata,
             coordinates=coordinates,
             coordinate_system=coordinate_system,
             detection_origin=detection_origin,
         )
 
+    def calculate_hash(
+        self, metadata: Optional[ElementMetadata] = None, other: Any = None
+    ) -> HashValue:
+        """Calculate the hash depending on what element data is available.
+
+        Args:
+            metadata - if provided, it will be included in the hash calculation.
+            other - any other data that should be included in the hash calculation.
+                Must implement __str__.
+
+        Returns:
+            HashValue - 128-bit hash value of the element.
+        """
+        if metadata is not None:
+            # Metadata supported by PDF and HTML files
+            data = f"{self.text}{metadata.page_number}{metadata.index_on_page}{other}"
+        else:
+            data = f"{self.text}{other}"
+
+        return HashValue(hashlib.sha256(data.encode()).hexdigest()[:32])
+
+    @property
+    def id(self) -> str:
+        if isinstance(self._id, (NoID, HashValue)):
+            # Ensure that hash is always up-to-date
+            self._id = self.calculate_hash(self.metadata)
+        return str(self._id)
```

Comment:
Here are some underlying domain-model concepts for `Element.id` which I've developed as I've studied these behaviors a little more closely:
### Basic design principles
- **Each `Element` assigns its own ID.** It is not intended for users to choose and assign ids to elements. The code for assigning IDs is internal to the element instance and is exercised on construction. An element deserialized from JSON receives the ID it was assigned by its original `Element` instance before serialization (as a `str` value).
- **Element IDs can be document-unique or globally unique.** An element can have an id based on a hash of its contents that is document-unique and deterministic, or it can have a globally-unique UUID which is not deterministic. Deterministic in this context means repartitioning the document will produce the same ids for all its elements. These are the only two options.
- **Element IDs are type `str`.** Regardless of the type of the `element_id` argument used to construct an element, `element.id` is type `str`. It can never be `None` and a UUID ID is resolved to its standard `str` representation, never the raw `uuid.UUID` object it may have come from. Internal storage is not specified by this and may be non-`str` types at the developer's discretion, but to a user, `element.id` is always a string.
### Construction-time ID behaviors
The `Element` constructor includes an `element_id` argument. This can accept three different values each producing a distinct ID behavior:
- `NoID` (default) - when the `element_id` argument is omitted or otherwise is an instance of `NoID`, the element is assigned a hash-based ID.
- `UUID` - an `element_id` argument that is an instance of `UUID()` (private "dummy" class, not `uuid.UUID`) causes the element to be assigned a UUID ID.
- `str` - a `str` element_id is assumed to be passed only by `unstructured` deserialization code when an element is deserialized from JSON. The element gets this same id initially assigned by the `Element` instance before it was serialized.

Not strictly a "construction-time" behavior, but an element can be assigned a `UUID` ID by calling `element.id_to_uuid()`. This is what the `@process_metadata` decorator does in response to the partitioning argument `unique_element_ids=True`.

Note that passing a `uuid.UUID` object as the `Element` constructor's `element_id` argument is not supported and will raise `ValueError`.

Also note that `Element` itself is a base class and should never be instantiated (in a partition element-stream at least). So such an instance does not (officially) have an `.id`.

Reply:
### `._id_type`

Reasoning with the premises stated above:
- An element implicitly has an `._id_type` attribute, conceptually taking one of the two values `HashValue` or `UUID` (or whatever actual values, this is just the gist).
- The value of this implicit `._id_type` is being inferred in the new `Text.id` property using `isinstance(self._id, (NoID, HashValue))`. Essentially the computation is:
	```python
	id_type = (
	    HashValue
	    if isinstance(self._id, (HashValue, NoID)
	    else UUID
	)
	```
- When `._id_type` is `HashValue`, the id is recomputed on each access, allowing content mutation to be reflected in the id. This is important because in some parts of the code the element is constructed in one step and the metadata added in a separate step.
- When `._id_type` is `UUID` the id is not recomputed.

Here, finally, is the rub. This implicit `._id_type` does not (currently) survive serialization. This means that a deserialized `Element` behaves differently than a "fresh" one that has never been serialized. I consider that a problem. It's way more than I think developers should have to discover for themselves and then be mindful of when debugging some weird bug.

I see three alternatives:
1. Make `Element._id_type` explicit, include it in the serialized JSON, and restore its value on deserialization.
2. Work out the id-type by the form of the string, like length, character set, dashes in certain spots maybe, etc. and restore it on deserialization (`isinstance(element_type, str)` on construction) to `HashValue` at that time rather than `str` such that it will trigger recomputation of the id on each access and behavior is not changed by round-tripping through JSON.
3. Accept that the behavior is way more complicated than we'll be willing or able to explain to users, leave it the way it is, document it carefully here in the code and hope for the best.

In principle I like option 1 but it seems over-engineered. Option 2 might be worthwhile if it's easy and reliable to distinguish a UUID from our SHA-hash. In any case I think we need to document all these quirks carefully (in docstrings and comments) while we understand it so the next developer doesn't have to spend the time we have working out how all this fits together :)

Reply:
all relevant points are summarized in overview.rst, if I overlooked something crucial, just let me know ;)